python/ray/tune/trial_executor.py

```diff
@@ -61,24 +60,24 @@ def stop_trial(self, trial, error=False, error_msg=None, stop_logger=True):
                                   "stop_trial() method")
 
     def restart_trial(self, trial, error_msg=None):
-        """Restarts the trial.
+        """Restarts or requeues the trial.
 
-        The state of the trial should restore from the last checkpoint.
+        The state of the trial should restore from the last checkpoint. Trial
+        is requeued if the cluster no longer has resources to accomodate it.
 
         Args:
             error_msg (str): Optional error message.
         """
-        try:
-            logger.info(
-                "Attempting to recover trial state from last checkpoint")
-            self.stop_trial(
-                trial, error=True, error_msg=error_msg, stop_logger=False)
-            trial.result_logger.flush()
+        self.stop_trial(
+            trial,
+            error=error_msg is not None,
+            error_msg=error_msg,
+            stop_logger=False)
+        trial.result_logger.flush()
+        if self.has_resources(trial.resources):
             self.start_trial(trial)
-        except Exception:
-            error_msg = traceback.format_exc()
-            logger.exception("Error recovering trial from checkpoint, abort.")
-            self.stop_trial(trial, error=True, error_msg=error_msg)
+        else:
+            trial.status = Trial.PENDING
```

Comment:
@joyyoj I removed the try-catch here because this method is only invoked by the `trial_runner`  and that already has a try-catch for this.